File: line_coding.py
"""
Name : Vedang Kamat
Roll No : 22B-ET-074
Class : BE, SEM VII
Branch : Electronics and Telecommunication Engineering
Goa College of Engineering, Farmagudi
"""

import matplotlib.pyplot as plt

# ====================================================
# Encoding Functions
# ====================================================

def ami_encode(bits):
    """AMI (Alternate Mark Inversion)"""
    encoded, prev = [], -1
    for b in bits:
        if b == 0:
            encoded.append(0)
        else:
            prev = 1 if prev == -1 else -1
            encoded.append(prev)
    return encoded

def b8zs_encode(bits):
    """B8ZS (Bipolar with 8-Zero Substitution)"""
    encoded, last_polarity, zero_count = [], -1, 0
    for b in bits:
        if b == 1:
            encoded.extend([0]*zero_count)
            zero_count = 0
            last_polarity = -last_polarity
            encoded.append(last_polarity)
        else:
            zero_count += 1
            if zero_count == 8:
                if last_polarity == 1:
                    encoded.extend([0,0,0,+1,-1,0,-1,+1])
                    last_polarity = +1
                else:
                    encoded.extend([0,0,0,-1,+1,0,+1,-1])
                    last_polarity = -1
                zero_count = 0
    encoded.extend([0]*zero_count)
    return encoded

def hdb3_encode(bits):
    """HDB3 (High-Density Bipolar-3)"""
    encoded, last_polarity, nonzero_count, zero_count = [], -1, 0, 0
    for b in bits:
        if b == 1:
            encoded.extend([0]*zero_count)
            zero_count = 0
            last_polarity = -last_polarity
            encoded.append(last_polarity)
            nonzero_count += 1
        else:
            zero_count += 1
            if zero_count == 4:
                if nonzero_count % 2 == 0:
                    b = 1 if last_polarity == -1 else -1
                    v = 1 if last_polarity == -1 else -1
                    encoded.extend([b,0,0,v])
                    last_polarity = -last_polarity
                    nonzero_count += 2
                else:
                    v = last_polarity
                    encoded.extend([0,0,0,v])
                    nonzero_count += 1
                zero_count = 0
    encoded.extend([0]*zero_count)
    return encoded

def manchester_encode(bits):
    """Manchester: 0 → low-high, 1 → high-low"""
    encoded = []
    for b in bits:
        encoded.extend([0,1] if b == 0 else [1,0])
    return encoded

def diff_manchester_encode(bits):
    """Differential Manchester"""
    encoded, prev = [], 1
    for b in bits:
        if b == 0:
            prev = 0 if prev == 1 else 1
        encoded.append(prev)
        prev = 0 if prev == 1 else 1
        encoded.append(prev)
    return encoded

def biphasem_encode(bits):
    """Biphase-M: transition at start; extra transition for 1"""
    encoded, prev = [], 0
    for b in bits:
        prev = 0 if prev == 1 else 1
        encoded.append(prev)
        if b == 1:
            prev = 0 if prev == 1 else 1
        encoded.append(prev)
    return encoded

def biphases_encode(bits):
    """Biphase-S: transition at start; extra transition for 0"""
    encoded, prev = [], 0
    for b in bits:
        prev = 0 if prev == 1 else 1
        encoded.append(prev)
        if b == 0:
            prev = 0 if prev == 1 else 1
        encoded.append(prev)
    return encoded

def nrzl_encode(bits):
    """Unipolar NRZ-L"""
    return [1 if b == 1 else 0 for b in bits]

def nrzm_encode(bits):
    """NRZ-M: 1 → invert, 0 → keep"""
    encoded, current = [], 1
    for b in bits:
        if b == 1:
            current = 0 if current == 1 else 1
        encoded.append(current)
    return encoded

def nrzs_encode(bits):
    """NRZ-S: 0 → invert, 1 → keep"""
    encoded, current = [], 0
    for b in bits:
        if b == 0:
            current = 0 if current == 1 else 1
        encoded.append(current)
    return encoded

def unipolar_rz(bits):
    """Unipolar RZ: 1 → [1,0], 0 → [0,0]"""
    encoded = []
    for b in bits:
        encoded.extend([1,0] if b == 1 else [0,0])
    return encoded


# ====================================================
# Main
# ====================================================

def main():
    encodings = {
        "1": ("Unipolar NRZ-L", nrzl_encode, "unipolar"),
        "2": ("Unipolar NRZ-M", nrzm_encode, "unipolar"),
        "3": ("Unipolar NRZ-S", nrzs_encode, "unipolar"),
        "4": ("Unipolar RZ", unipolar_rz, "rz"),
        "5": ("Manchester", manchester_encode, "rz"),
        "6": ("Differential Manchester", diff_manchester_encode, "rz"),
        "7": ("Biphase-M", biphasem_encode, "rz"),
        "8": ("Biphase-S", biphases_encode, "rz"),
        "9": ("AMI", ami_encode, "bipolar"),
        "10": ("HDB3", hdb3_encode, "bipolar"),
        "11": ("B8ZS", b8zs_encode, "bipolar"),
    }

    # Default bitstream
    default_bits = "100000000100001"
    print("\nAvailable Encodings:")
    for k,v in encodings.items():
        print(f"{k}. {v[0]}")

    choice = input("\nChoose encoding [1-11]: ").strip()
    if choice not in encodings:
        print("Invalid choice.")
        return

    name, func, etype = encodings[choice]

    # Input bit sequence
    input_str = input(f"Enter bit sequence [default: {default_bits}]: ").strip()
    if input_str == "":
        input_str = default_bits

    try:
        bits = [int(ch) for ch in input_str]
        if any(b not in (0,1) for b in bits):
            raise ValueError
    except ValueError:
        print("Invalid input! Only 0/1 allowed.")
        return

    n = len(bits)
    encoded = func(bits)

    # Extend input
    x_bits = list(range(n+1))
    bits_ext = bits + [bits[-1]]

    # Handle encoding axis
    if etype == "rz":  # double resolution
        x_encoded = [i*0.5 for i in range(len(encoded)+1)]
    else:
        x_encoded = list(range(len(encoded)+1))
    encoded_ext = encoded + [encoded[-1]]

    # Plotting
    plt.figure(figsize=(12,6))

    # Input
    plt.subplot(2,1,1)
    plt.step(x_bits, bits_ext, where="post", color="black", linewidth=2)
    plt.fill_between(x_bits, bits_ext, step="post", alpha=0.3, color="black")
    plt.ylim(-0.2, 1.2)
    plt.yticks([0,1], ["0","1"])
    plt.xticks(range(n+1))
    plt.xlim(0, n)
    plt.grid(True, linestyle="--", linewidth=0.7, color="gray")
    plt.title("Input Bitstream")

    # Encoded
    plt.subplot(2,1,2)
    plt.step(x_encoded, encoded_ext, where="post", color="black", linewidth=2)
    plt.fill_between(x_encoded, encoded_ext, step="post", alpha=0.3, color="black")
    
    if etype == "bipolar":
        plt.ylim(-1.2, 1.2)
        plt.yticks([-1,0,1], ["-V","0","+V"])
    else:
        plt.ylim(-0.2, 1.2)
        plt.yticks([0,1], ["0","+V"])

    plt.xticks(range(n+1))
    plt.xlim(0, n)
    plt.grid(True, linestyle="--", linewidth=0.7, color="gray")
    plt.title(f"{name} Encoded Signal")

    plt.xlabel("Bit Index")
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()

File: line_coding.c
/*
 Name   : Vedang Kamat
 Roll No: 22B-ET-074
 Class  : BE, SEM VII
 Branch : Electronics and Telecommunication Engineering
 Goa College of Engineering, Farmagudi
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// ================= Encoding Functions =================

// NRZ-L encoding
void nrzl_encode(int *b, int n) {
    printf("NRZ-L             : ");
    for (int i = 0; i < n; i++)
        printf(b[i] ? "+V | " : "0 | ");
    printf("\n");
}

// NRZ-M encoding
void nrzm_encode(int *b, int n) {
    printf("NRZ-M             : ");
    int current = 1;
    for (int i = 0; i < n; i++) {
        if (b[i] == 1) current = !current;
        printf(current ? "+V | " : "0 | ");
    }
    printf("\n");
}

// NRZ-S encoding
void nrzs_encode(int *b, int n) {
    printf("NRZ-S             : ");
    int current = 0;
    for (int i = 0; i < n; i++) {
        if (b[i] == 0) current = !current;
        printf(current ? "+V | " : "0 | ");
    }
    printf("\n");
}

// Manchester encoding
void manchester_encode(int *b, int n) {
    printf("Manchester        : ");
    for (int i = 0; i < n; i++) {
        int clk = 1;
        printf((!(b[i] ^ clk)) ? "+V " : "0 ");
        clk = 0;
        printf((!(b[i] ^ clk)) ? "+V | " : "0 | ");
    }
    printf("\n");
}

// Differential Manchester encoding
void diff_manchester_encode(int *b, int n) {
    printf("Diff Manchester   : ");
    int prev = 1;
    for (int i = 0; i < n; i++) {
        if (b[i] == 0) {
            prev = !prev;
            printf(prev ? "+V " : "0 ");
            prev = !prev;
            printf(prev ? "+V | " : "0 | ");
        } else {
            printf(prev ? "+V " : "0 ");
            prev = !prev;
            printf(prev ? "+V | " : "0 | ");
        }
    }
    printf("\n");
}

// Biphase-M encoding
void biphasem_encode(int *b, int n) {
    printf("Biphase-M         : ");
    int prev = 0;
    for (int i = 0; i < n; i++) {
        prev = !prev;
        printf(prev ? "+V " : "0 ");
        if (b[i] == 1) prev = !prev;
        printf(prev ? "+V | " : "0 | ");
    }
    printf("\n");
}

// Biphase-S encoding
void biphases_encode(int *b, int n) {
    printf("Biphase-S         : ");
    int prev = 0;
    for (int i = 0; i < n; i++) {
        prev = !prev;
        printf(prev ? "+V " : "0 ");
        if (b[i] == 0) prev = !prev;
        printf(prev ? "+V | " : "0 | ");
    }
    printf("\n");
}

// AMI encoding
void ami_encode(int *b, int n) {
    printf("AMI               : ");
    int prev = 0;
    for (int i = 0; i < n; i++) {
        if (b[i] == 1) {
            prev = !prev;
            printf(prev ? "+V | " : "-V | ");
        } else {
            printf("0 | ");
        }
    }
    printf("\n");
}

// HDB3 encoding
void hdb3_encode(int *b, int n) {
    printf("HDB3              : ");
    int last_polarity = -1;
    int nonzero_count = 0;
    int zero_count = 0;

    for (int i = 0; i < n; i++) {
        if (b[i] == 1) {
            for (int z = 0; z < zero_count; z++) printf("0 | ");
            zero_count = 0;

            last_polarity = -last_polarity;
            printf(last_polarity == 1 ? "+V | " : "-V | ");
            nonzero_count++;
        } else {
            zero_count++;
            if (zero_count == 4) {
                if (nonzero_count % 2 == 0) {
                    printf("%s | 0 | 0 | %s | ",
                           last_polarity == -1 ? "+B" : "-B",
                           last_polarity == -1 ? "+V" : "-V");
                    last_polarity = -last_polarity;
                    nonzero_count += 2;
                } else {
                    printf("0 | 0 | 0 | %s | ", last_polarity == 1 ? "+V" : "-V");
                    nonzero_count++;
                }
                zero_count = 0;
            }
        }
    }

    for (int z = 0; z < zero_count; z++) printf("0 | ");
    printf("\n");
}

// B8ZS encoding
void B8ZS_encode(int *b, int n) {
    printf("B8ZS              : ");
    int last_polarity = -1;
    int zero_count = 0;

    for (int i = 0; i < n; i++) {
        if (b[i] == 1) {
            for (int z = 0; z < zero_count; z++) printf("0 | ");
            zero_count = 0;

            last_polarity = -last_polarity;
            printf(last_polarity == 1 ? "+V | " : "-V | ");
        } else {
            zero_count++;
            if (zero_count == 8) {
                if (last_polarity == 1)
                    printf("0 | 0 | 0 | +V | -B | 0 | -V | +B | "), last_polarity = +1;
                else
                    printf("0 | 0 | 0 | -V | +B | 0 | +V | -B | "), last_polarity = -1;
                zero_count = 0;
            }
        }
    }

    for (int z = 0; z < zero_count; z++) printf("0 ");
    printf("\n");
}

// Unipolar RZ encoding
void unipolar_rz(int *b, int n) {
    printf("Unipolar RZ       : ");
    for (int i = 0; i < n; i++) {
        if (b[i] == 1)
            printf("+V 0 | ");
        else
            printf("0 0 | ");
    }
    printf("\n");
}

// ================= Main Program =================

int main() {
    char *input;
    int *b, n;
    size_t max_len = 100;
    const char *default_bits = "100000000100001";  // default sequence

    input = malloc((max_len + 1) * sizeof(char));
    if (!input) {
        printf("Memory allocation failed!\n");
        return 1;
    }

    printf("\nEnter the bit sequence (default = %s): ", default_bits);
    if (!fgets(input, max_len, stdin)) {
        printf("Input error!\n");
        free(input);
        return 1;
    }

    // Remove trailing newline if present
    input[strcspn(input, "\n")] = '\0';

    // If input is empty, use default
    if (strlen(input) == 0) {
        strcpy(input, default_bits);
    }

    n = strlen(input);
    b = malloc(n * sizeof(int));
    if (!b) {
        printf("Memory allocation failed!\n");
        free(input);
        return 1;
    }

    for (int i = 0; i < n; i++) {
        if (input[i] == '0') b[i] = 0;
        else if (input[i] == '1') b[i] = 1;
        else {
            printf("Invalid character detected!\n");
            free(b);
            free(input);
            return 1;
        }
    }

    // Print all encodings
    printf("\n=== Line Coding Schemes ===\n");
    nrzl_encode(b, n);
    nrzm_encode(b, n);
    nrzs_encode(b, n);
    manchester_encode(b, n);
    diff_manchester_encode(b, n);
    biphasem_encode(b, n);
    biphases_encode(b, n);
    ami_encode(b, n);
    hdb3_encode(b, n);
    B8ZS_encode(b, n);
    unipolar_rz(b, n);

    free(b);
    free(input);
    printf("\n");
    return 0;
}


